feat(blob|state): Blob submission metrics

gupadhyaya opened the pull request:
# Add Comprehensive Metrics for Blob Submission Path

## Overview

This PR adds comprehensive OpenTelemetry metrics to track performance and behavior of the blob submission path in both the `blob` and `state` modules. The metrics provide detailed insights into submission timing, error rates, gas estimation performance, and resource usage.

## What's Added

### Blob Module Metrics (`blob/metrics.go`)
- **Retrieval tracking**: Get operations, timing, and error rates  
- **Proof tracking**: Proof generation timing and success rates
- **Error categorization**: Detailed error tracking with context

**Note**: Submission metrics are handled by the state module to avoid duplication.

### State Module Metrics (`state/metrics.go`)
- **PFB submission tracking**: PayForBlob transaction metrics with gas details
- **Gas estimation metrics**: Timing and success rates for gas calculations
- **Gas price estimation**: Performance tracking for price estimation
- **Account query metrics**: Account lookup timing and error rates

## Key Metrics

### Blob Metrics
```
blob_retrieval_total
blob_retrieval_duration_seconds
blob_proof_total
blob_proof_duration_seconds
```

**Note**: Blob submission metrics are tracked in the state module to avoid duplication.

### State Metrics
```
state_pfb_submission_total{blob_count, total_size_bytes, gas_price_utia}
state_pfb_submission_duration_seconds
state_pfb_gas_estimation_duration_seconds
state_pfb_gas_price_estimation
state_gas_estimation_total
state_account_query_total
```

## How to Use

### 1. Enable Metrics in Node Configuration

```bash
# Start node with OTLP metrics enabled
./celestia bridge start --metrics --metrics.endpoint localhost:4318 --metrics.tls=false
./celestia light start --metrics --metrics.endpoint localhost:4318 --metrics.tls=false
```

### 2. Set Up OTLP Collector (Required for Grafana)

Create `otel-collector-config.yaml`:
```yaml
receivers:
  otlp:
    protocols:
      http:
        endpoint: 0.0.0.0:4318

exporters:
  prometheus:
    endpoint: "0.0.0.0:8889"
    namespace: "celestia"

service:
  pipelines:
    metrics:
      receivers: [otlp]
      exporters: [prometheus]
```

Run OTLP collector:
```bash
docker run -p 4318:4318 -p 8889:8889 -v $(pwd)/otel-collector-config.yaml:/etc/otel-collector-config.yaml otel/opentelemetry-collector-contrib:latest --config=/etc/otel-collector-config.yaml
```

### 3. Query Metrics Endpoint

```bash
# Get all metrics from OTLP collector
curl http://localhost:8889/metrics

# Filter blob metrics
curl http://localhost:8889/metrics | grep "blob_"

# Filter state metrics  
curl http://localhost:8889/metrics | grep "state_"
```

### 4. Monitor Specific Operations

```bash
# Monitor blob retrieval operations
curl -s http://localhost:8889/metrics | grep "blob_retrieval_total"

# Monitor blob proof generation
curl -s http://localhost:8889/metrics | grep "blob_proof_total"

# Monitor gas estimation performance
curl -s http://localhost:8889/metrics | grep "state_pfb_gas_estimation"
```

### 5. Prometheus Integration

Add to your `prometheus.yml`:
```yaml
scrape_configs:
  - job_name: 'celestia-node-otlp'
    static_configs:
      - targets: ['localhost:8889']  # OTLP collector's Prometheus endpoint
    metrics_path: '/metrics'
    scrape_interval: 15s
```

### 6. Grafana Dashboard

Configure Grafana to use Prometheus datasource pointing to `http://localhost:8889`

Create dashboards to visualize:
- Blob retrieval and proof generation performance
- State PFB submission performance  
- Error rates and types
- Gas estimation metrics
- Account query performance

### 7. Architecture Overview

```
Celestia Node (OpenTelemetry) → OTLP Collector → Prometheus Format → Grafana
```

- **Celestia Node**: Exports OpenTelemetry metrics to OTLP collector
- **OTLP Collector**: Converts OpenTelemetry metrics to Prometheus format
- **Grafana**: Visualizes metrics from Prometheus endpoint

## Example Metrics Output

```
# State PFB metrics (from OTLP collector)
celestia_pfb_count_total{instance="12D3KooW...",job="test/Bridge",label1="value1",otel_scope_name="state"} 5
celestia_last_pfb_timestamp_total{instance="12D3KooW...",job="test/Bridge",label1="value1",otel_scope_name="state"} 1.758100573267e+12

# Blob metrics
celestia_blob_retrieval_observable_total{instance="12D3KooW...",job="test/Bridge",label1="value1",otel_scope_name="blob"} 0
celestia_blob_proof_observable_total{instance="12D3KooW...",job="test/Bridge",label1="value1",otel_scope_name="blob"} 0

# Discovery metrics
celestia_archival_discovery_amount_of_peers{instance="12D3KooW...",job="test/Bridge",label1="value1",otel_scope_name="share_discovery"} 0
celestia_archival_discovery_find_peers_result_total{enough_peers="false",instance="12D3KooW...",job="test/Bridge",label1="value1",otel_scope_name="share_discovery"} 65
```

**Note**: Metrics are prefixed with `celestia_` and include OpenTelemetry scope information when exported through the OTLP collector.

Fixes #4538

gupadhyaya committed:
add blob submission metrics

gupadhyaya committed:
fix tests

gupadhyaya committed:
fix build errors

gupadhyaya committed:
fix api

gupadhyaya committed:
shorten

gupadhyaya committed:
Merge branch 'main' into blob_submission_metrics

Wondertan commented on blob/metrics.go:
```diff
@@ -0,0 +1,297 @@
+// Metrics tracks blob-related metrics
+type Metrics struct {
+	// Submission metrics
+	submissionCounter   metric.Int64Counter
+	submissionDuration  metric.Float64Histogram
+	submissionErrors    metric.Int64Counter
+	submissionBlobCount metric.Int64Counter
+	submissionBlobSize  metric.Int64Counter
```
Duplicates state module metrics. Do we want this? Doesn't seem like

Wondertan reviewed (commented):
Solid. Mainly one comment about metrics duplication. Also wonder if you tested this on Graphana, so we don't accidentally get nil ptr or a metrics that doesn work

gupadhyaya committed:
remove duplicates

gupadhyaya committed:
fix format issue

gupadhyaya committed:
Fix bitswap compilation issues and enable metrics in tastora framework

- Remove unsupported broadcast control options from bitswap
- Add missing HashOnRead method to BlockstoreWithMetrics
- Enable metrics flags in tastora framework for bridge and light nodes

gupadhyaya committed:
Remove HashOnRead method from BlockstoreWithMetrics

The HashOnRead method is not part of the current boxo Blockstore interface

gupadhyaya committed:
Update tastora framework to use local Docker image with metrics

- Change nodeImage to use local celestia-node-local image
- Update defaultNodeTag to use our commit 6bc23167 with metrics fixes

gupadhyaya committed:
Remove interface compliance check for BlockstoreWithMetrics

The interface check was failing because HashOnRead method was removed from the Blockstore interface

gupadhyaya committed:
fix coreaccessor with new metrics instead of nil, fx.invoke issue

gupadhyaya committed:
fix blob metrics

gupadhyaya committed:
go.mod changes

gupadhyaya committed:
Merge branch 'main' into blob_submission_metrics

gupadhyaya committed:
go mod tidy

gupadhyaya committed:
blob metrics fixes

gupadhyaya committed:
fix

gupadhyaya committed:
remove unnecessary method

gupadhyaya commented on nodebuilder/tests/tastora/framework.go:
```diff
@@ -360,7 +360,7 @@ func (f *Framework) startBridgeNode(ctx context.Context, chain tastoratypes.Chai
 	err = bridgeNode.Start(ctx,
 		tastoratypes.WithChainID(testChainID),
-		tastoratypes.WithAdditionalStartArguments("--p2p.network", testChainID, "--core.ip", hostname, "--rpc.addr", "0.0.0.0"),
+		tastoratypes.WithAdditionalStartArguments("--p2p.network", testChainID, "--core.ip", hostname, "--rpc.addr", "0.0.0.0", "--metrics", "--p2p.metrics", "--metrics.endpoint", "host.docker.internal:4318", "--metrics.tls=false"),
```
correct.

gupadhyaya committed:
remove debug stuff

gupadhyaya committed:
WithMetrics

Wondertan commented on state/metrics.go:
```diff
@@ -2,15 +2,246 @@ package state
+	totalAccountQueryErrors       atomic.Int64
+}
+
+func WithMetrics(lc fx.Lifecycle, ca *CoreAccessor) (*Metrics, error) {
```
You don't need to provide nil Metrics if the Metrics struct is not part of the dependency graph.

gupadhyaya committed:
remove passing nil metrics

gupadhyaya committed:
private metrics, stop method

gupadhyaya commented on state/metrics.go:
```diff
@@ -2,15 +2,247 @@ package state
+}
+
+// WithMetrics initializes metrics for the CoreAccessor
+func (ca *CoreAccessor) WithMetrics(lc fx.Lifecycle) (*Metrics, error) {
```
added

gupadhyaya replied:
removed fx and added stop method to unregister

gupadhyaya committed:
cleanup

gupadhyaya committed:
Merge branch 'main' into blob_submission_metrics

Wondertan commented on state/core_access.go:
```diff
@@ -501,13 +543,53 @@ func (ca *CoreAccessor) PayForBlobCount() int64 {
 }
 
+// SetMetrics updates the metrics for the CoreAccessor
+func (ca *CoreAccessor) SetMetrics(metrics *metrics) {
```
Lets remove this from public api as its cant and not going to be used. I think we can directly assign to metrics field in place without the method and locking.

gupadhyaya replied:
done

gupadhyaya committed:
blob metrics changes

gupadhyaya commented on blob/metrics.go:
```diff
@@ -0,0 +1,196 @@
+}
+
+// WithMetrics registers blob metrics
+func WithMetrics(lc fx.Lifecycle) (*Metrics, error) {
```
done.

Wondertan reviewed (commented):
almost there

gupadhyaya committed:
remove nil checks

gupadhyaya committed:
get rid of duplicate legacy state metrics

gupadhyaya committed:
optimizing metrics recording

gupadhyaya committed:
error out if metrics fail

gupadhyaya committed:
named error

gupadhyaya committed:
skip resp name

gupadhyaya committed:
error as bool/enum to avoid cardinality explosion

gupadhyaya committed:
golangci-lint fix

gupadhyaya committed:
unify error counting

Wondertan commented on blob/metrics.go:
```diff
@@ -0,0 +1,188 @@
+		errorType := "unknown"
+		switch {
+		case errors.Is(err, ErrBlobNotFound):
+			errorType = "not_found"
```
Can we make these magic strings to be defined as constants?

Wondertan commented on state/metrics.go:
```diff
@@ -2,41 +2,402 @@ package state
+		attribute.Int("blob_count", blobCount),
+		attribute.Int64("total_size_bytes", totalSize),
+		attribute.Float64("gas_price_utia", gasPrice),
+		attribute.Bool("success", err == nil),
```
Ideally we would define some error types here too, but that can be done as follow up once we enumarate all of them

gupadhyaya committed:
remove magic strings

gupadhyaya committed:
golangci-lint fixes

gupadhyaya committed:
further fix golangci-lint

gupadhyaya committed:
fixing docker security issue

gupadhyaya committed:
Revert "fixing docker security issue"

This reverts commit 533f70ecd9d7fdad35684df02973ba22c7305382.

renaynay reviewed (changes requested):
* pls make sure go mods for our internal deps (celestiaorg) match what's on main branch (like don't update go-square)
* I'd prefer to do metric observations (time it took) in the helper methods themselves instead of inline in the submission code path as the submission code path is already extremely long and convoluted

gupadhyaya committed:
Merge origin/main into blob_submission_metrics

Resolved conflicts in:
- nodebuilder/header/config.go: Fixed return value in trustedPeers function
- share/root.go: Fixed return value in RowsWithNamespace function
- state/core_access.go: Kept metrics-related code from HEAD
- state/core_access_test.go: Kept metrics parameter in NewCoreAccessor
- nodebuilder/tests/tastora/go.mod: Used origin/main versions
- nodebuilder/tests/tastora/go.sum: Used origin/main versions

gupadhyaya committed:
Fix compilation error after merge

- Fixed estimateGasForBlobs function call to match new signature
- Properly declared author variable to avoid redeclaration
- Maintained metrics functionality while fixing function signature changes

gupadhyaya committed:
Apply goimports-reviser formatting to fix import organization

- Applied goimports-reviser formatting to all Go files
- Fixed import grouping and ordering according to project standards
- This addresses the lint-imports failures after the merge

gupadhyaya committed:
Fix import ordering in state/core_access.go

- Reordered imports to match goimports-reviser expectations
- Grouped imports: standard library, third-party, external project, local project
- This resolves the lint-imports failure

gupadhyaya committed:
Fix remaining issues after merge

- Fix duplicate // indirect comment in go.mod
- Fix compilation error in state/core_access_test.go where accounts[0].Name was called on a string

gupadhyaya committed:
minor

gupadhyaya committed:
go mod tidy

renaynay reviewed (commented):
* Lint errors that need resolving
* thank you for adding grafana dashboard, I UTACK that
* might need to be rebased
* bunch of root level files that do not belong at root
* some of my comments re necessity of some of the metrics not addressed - pls let's not overbloat w unnecessary metrics.

gupadhyaya committed:
remove unnecessary metrics

gupadhyaya committed:
Merge branch 'main' into blob_submission_metrics

gupadhyaya committed:
Remove accidentally committed grafana and metrics files

gupadhyaya committed:
Fix TestListenerWithWrongChainRPC cleanup issue

gupadhyaya committed:
Merge remote-tracking branch 'origin/main' into blob_submission_metrics

gupadhyaya committed:
fix

gupadhyaya committed:
minor fix

gupadhyaya commented:
closing in favor of https://github.com/celestiaorg/celestia-node/pull/4664
